File: Data Structures & Algorithms/max-water-container/submission-3.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1, res = 0;
        while (l < r) {
            res = max(min(heights[l], heights[r]) * (r-l), res);
            if (heights[l] < heights[r]) l += 1;
            else r -= 1;
        }
        return res;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-3.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        

        n = len(heights)
        left, right = [0]* n, [0] * n
        left[0] = 0
        right[-1] = 0
        for i in range(1, n):
            left[i] = max(left[i-1], heights[i-1])
        for i in range(n-2, -1, -1):
            right[i] = max(right[i+1], heights[i+1])
        res = 0
        for i in range(n):
            res += min(left[i], right[i]) - heights[i] if min(left[i], right[i]) - heights[i] > 0 else 0 
        return res